pythia/models/butd.py

```diff
@@ -108,52 +109,92 @@ def get_data_t(self, t, data, batch_size_t, prev_output):
 
         return data, batch_size_t
 
-    def forward(self, sample_list):
-        # Stores the output probabilites. Not used if beam_search inference
-        scores = sample_list.answers.new_ones(
-            (
-                sample_list.answers.size(0),
-                self.text_processor.max_length,
-                self.vocab_size,
-            ),
-            dtype=torch.float,
-        )
+    def forward_greedy(self, sample_list, scores):
 
-        # For beam search inference. Currently beam seach for BUTD works only 
-        # with batch_size = 1 and should be used with run_type inference only.
-        # TODO : Implement batch beam search
-        if self.config["inference"]["type"] == "beam_search":
-            beam_search = BeamSearch(
-                self.vocab, self.config["inference"]["params"]["beam_length"]
+        batch_size = sample_list.image_feature_0.size(0)
+        data, sample_list, timesteps = self.prepare_data(sample_list, batch_size)
+        output = None
+        batch_size_t = batch_size
+        for t in range(timesteps):
+            data, batch_size_t = self.get_data_t(t, data, batch_size_t, output)
+            pi_t = data["texts"][:, t].unsqueeze(-1)
+            embedding = self.word_embedding(pi_t)
+            attention_feature, _ = self.process_feature_embedding(
+                "image", sample_list, embedding[:, 0, :], batch_size_t=batch_size_t
             )
-            sample_list = beam_search.init_batch(sample_list)
+            output = self.classifier(attention_feature)
+            scores[:batch_size_t, t] = output
+
+        model_output = {"scores": scores}
+        return model_output
```

Comment:
Maybe I am missing something, but shouldn't `forward_greedy` return the final caption too? `forward_beam_search` and `forward_nucleus_sampling` seems to be returning an additional key in the dictionary: `captions`.

Reply:
The current version of butd.py only returns scores for greedy. I guess for greedy the caption generation is being handled somewhere else.

Reply:
@dexter1691 you are right about this. I checked and greedy captioning shows error as it doesn't have 'captions' key in the dictionary. 

A short fix can be to reuse beam search module with beam_size = 1 to get captions. What do you think? Tagging @vedanuj (author).

Reply:
So, here is a better way to do this:
Create a base class with `TextDecoder` which `BeamSearch` and `NucleusSampling` inherit. Greedy can just be beam search initialized with beam size = 1. You don't need separate forward for it.

Now, you don't need separate forward for all 3. Try to converge them in a pattern using encapsulation so that a single forward can be written for all 3. 

Seeing the code I see a lot of code getting repeated. Overall logic will remain same and all of the custom logic will go inside the decoders you implement. 

Let me know if this isn't clear to you. We can discuss more.

Reply:
Add this to modules decoders afterwards.

Reply:
I understand the base class approach. Should be pretty straightforward. Will update once done.

Reply:
I've made a base class that contains all the common code but I'm not able to figure out how to converge 3 forwards into one. An object of either BeamSearch or NucleusSampling needs to get created. A common forward can only do it through switch case or if-else. Even then, we can reduce a lot of redundancy. Should I make a common forward having conditional statements?